lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,474 @@
+		numTxnBytes, err := ReadUvarint(rr)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsMetadata.FromBytes: Problem reading number of bytes in transaction")
+		}
+
+		// Allocate memory for the transaction bytes to be read into.
+		txnBytes, err := SafeMakeSliceWithLength[byte](numTxnBytes)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsMetadata.FromBytes: Problem allocating bytes for transaction")
+		}
+
+		// Read the transaction into the txnBytes memory buffer.
+		if _, err = io.ReadFull(rr, txnBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsMetadata.FromBytes: Problem reading bytes for transaction")
+		}
+
+		// Convert the txnBytes buffer to a MsgDeSoTxn struct.
+		if err = txnData.Txns[ii].FromBytes(txnBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsMetadata.FromBytes: Problem parsing transaction bytes")
+		}
+	}
+	return nil
+}
+
+func (txnData *AtomicTxnsMetadata) New() DeSoTxnMetadata {
+	return &AtomicTxnsMetadata{}
+}
+
+//
+// HELPER FUNCTIONS: MsgDeSoTxn
+//
+
+func (msg *MsgDeSoTxn) IsAtomicTxnsInnerTxn() bool {
+	// An atomic transaction is qualified by the existence of the NextAtomicTxnPreHash
+	// and PreviousAtomicTxnPreHash keys in the ExtraData map.
+	if _, keyExists := msg.ExtraData[NextAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	if _, keyExists := msg.ExtraData[PreviousAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	return true
+}
+
+func (msg *MsgDeSoTxn) AtomicHash() (*BlockHash, error) {
+	// Create a duplicate of the transaction to ensure we don't edit the existing transaction.
+	msgDuplicate, err := msg.Copy()
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: Cannot create duplicate transaction")
+	}
+
+	// Sanity check that the transaction includes the necessary extra data to be included in an atomic transaction.
+	if !msgDuplicate.IsAtomicTxnsInnerTxn() {
+		return nil, errors.New("MsgDeSoTxn.AtomicHash: Cannot compute atomic hash on non-atomic transaction")
+	}
+
+	// Delete the NextAtomicTxnPreHash and PreviousAtomicTxnPreHash from the ExtraData map.
+	delete(msgDuplicate.ExtraData, NextAtomicTxnPreHash)
+	delete(msgDuplicate.ExtraData, PreviousAtomicTxnPreHash)
+
+	// Convert the transaction to bytes but do NOT encode the transaction signature.
+	preSignature := true
+	txBytes, err := msgDuplicate.ToBytes(preSignature)
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: cannot convert modified transaction to bytes")
+	}
+
+	// Return the SHA256 double hash of the resulting bytes.
+	return Sha256DoubleHash(txBytes), nil
+}
+
+//
+// Connect and Disconnect Atomic Txn Logic
+//
+
+func (bav *UtxoView) _connectAtomicTxns(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	blockTimestampNanoSecs int64,
+	verifySignature bool,
+) (
+	_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error,
+) {
+	var utxoOpsForTxn []*UtxoOperation
+
```

Comment:
Should we cap how many transactions an atomic txns wrapper can contain?

Reply:
Nah I think max txn size should cover it.